### PR TITLE
Adding DisplayPlane Surface support for object tracking

### PR DESCRIPTION
**Pull request by pbantolas**
Having no WSI was not handled in the layer generation scripts. Added a
new DisplayServer named 'Null'.

Added a missing layer function for vkCreateDisplayPlaneSurfaceKHR. This
fixes errors when using a NULL window system reporting an invalid
VkSurface object.

### CODE REVIEW

**Comment by tobine** [comment]
Ian, can you give this a look and if you need further review of codegen stuff when you're done just assign to me.

**Comment by ianelliottus** [comment]
I apologize that I didn't see this pull request until now.  I would like more context about what is desired with these changes.  For example, there is no concept of a "DisplayServer" in the swapchain layer, and yet it adds a "NULL" one.  Also, this pull request seems to partially add support for the VK_KHR_display extension (not currently supported by the swapchain validation layer).  I say partially because only for one function of that extension is added.  Yet, the validation code is checking whether the VK_KHR_surface extension is enabled before using that function.

What is it that you are trying to validate?  Do you have an implementation of the VK_KHR_display and VK_KHR_display_swapchain extension?

**Comment by ianelliottus** [comment]
@pbantolas, there's action required by you.  Please respond to my questions.  Thanks!

**Comment by pbantolas** [comment]
@ianelliottus You are right, I have a typo in my commit for the vk-generate.py file - I'll commit a fix.

Do you think we should be handling the rest of the VK_KHR_display extension functions in the loader? They don't seem to need any special implementation but would like your opinion on that.

**Comment by ianelliottus** [comment]
Sorry to not get back to this for a while!

> Do you think we should be handling the rest of the VK_KHR_display extension functions in the loader?

I believe that LunarG added those functions a few weeks back.  Please look in "loader/wsi.c".

> They don't seem to need any special implementation but would like your opinion on that.

Yes, all instance extensions need loader support.  In particular, any function that dispatches from a VkInstance or VkPhysicalDevice requires both a "trampoline" and "terminator" function i nthe loader.  Please see: https://github.com/KhronosGroup/Vulkan-LoaderAndValidationLayers/blob/master/loader/LoaderAndLayerInterface.md.

Different question: do you know of an ICD that implements the VK_KHR_display and VK_KHR_display_swapchain extensions?  The validation code was deferred for those extensions because no implementations existed.  It will be much easier to write and test validation layer code once there's an ICD to try it with.

**Comment by michaelworcester** [comment]
Petros works for Imagination, and we have an ICD that exposes those extensions, which is why we're keen to get them supported.

**Comment by ianelliottus** [comment]
Thanks for your reply @michaelworcester!  I'll add this to the agenda for the WSI/System Integration meeting that starts at 9AM PDT.

**Comment by jon-lunarg** [comment]
Michael or Petros:
A couple questions.  What OS  does your ICD that supports these extension run on?  If on  WIndows or Ubuntu/Fedora where can we get it? Also the display_swapchain extension by the letter of the law needs loader support since the extension entry points are suppose to be exported. Would you guys be interested in adding that support to the loader in a Pull Request?

File: CMakeLists.txt [hunk 1]
```diff
@@ -22,6 +22,7 @@ elseif(CMAKE_SYSTEM_NAME STREQUAL "Linux")
     #         Mir support is stubbed in but unimplemented and untested.
     option(BUILD_WSI_XCB_SUPPORT "Build XCB WSI support" ON)
     option(BUILD_WSI_XLIB_SUPPORT "Build Xlib WSI support" ON)
+    option(BUILD_WSI_DISPLAY_SUPPORT "Build DisplayPlane surface WSI support" ON)
```
Comment (jon-lunarg) [inline CMakeLists.txt]: There is no platform ifdef in vulkan.h for the KHR_display or KHR_display_swapchain extensions. Can't these just be always built?  These extensions don't include any platform specifc headers like the Xlib/Xcb/Mir/Wayland/Win32
extensions do.

File: layers/swapchain.cpp [hunk 2]
```diff
@@ -863,6 +865,59 @@ VK_LAYER_EXPORT VKAPI_ATTR VkBool32 VKAPI_CALL vkGetPhysicalDeviceXlibPresentati
 }
 #endif // VK_USE_PLATFORM_XLIB_KHR
 
+VK_LAYER_EXPORT VKAPI_ATTR VkResult VKAPI_CALL
+vkCreateDisplayPlaneSurfaceKHR(VkInstance instance, const VkDisplaySurfaceCreateInfoKHR *pCreateInfo, const VkAllocationCallbacks *pAllocator,
+                       VkSurfaceKHR *pSurface) {
+    VkResult result = VK_SUCCESS;
+    bool skipCall = false;
+    layer_data *my_data = get_my_data_ptr(get_dispatch_key(instance), layer_data_map);
+    loader_platform_thread_lock_mutex(&globalLock);
```
Comment (jon-lunarg) [inline layers/swapchain.cpp]: the layers are now using STL unique_lock rather than loader_platform_thread_lock_mutex.

File: vk-generate.py [hunk 3]
```diff
@@ -248,7 +248,8 @@ def main():
             "Xcb",
             "Xlib",
             "Wayland",
-            "Mir"
+            "Mir",
+            "Display"
     }
```
Comment (jon-lunarg) [inline vk-generate.py]: I am not sure why this is needed to add this "Display" option?  I think the generated layers should just always 
support theses extension entrypoints.

**Comment by ianelliottus** [comment]
Hi @pbantolas and @michaelworcester.  I spoke with Jon Ashburn of LunarG, and he's going to help you flesh out this pull request, to handle the extensions.  Best wishes!

**Comment by michaelworcester** [comment]
@jon-lunarg 

> Michael or Petros:
> A couple questions. What OS does your ICD that supports these extension run on? If on WIndows or Ubuntu/Fedora where can we get it?

Linux, but as it's really an internal development/debug platform I don't think we can ship the hardware/driver externally, sorry. :(

> Also the display_swapchain extension by the letter of the law needs loader support since the extension entry points are suppose to be exported. Would you guys be interested in adding that support to the loader in a Pull Request?

We don't actually support this at the moment but adding support to the loader should be fine, let me take a look and get back to you.

Cheers for the feedback, @pbantolas is currently fixing all specific issues raised!

File: vk-layer-generate.py [hunk 4]
```diff
@@ -1383,14 +1384,15 @@ def generate_body(self):
                                    'vkGetPhysicalDeviceSurfaceFormatsKHR',
                                    'vkGetPhysicalDeviceSurfacePresentModesKHR',
                                    'vkCreateAndroidSurfaceKHR'])]
-        elif self.wsi == 'Xcb' or self.wsi == 'Xlib' or self.wsi == 'Wayland' or self.wsi == 'Mir':
+        elif self.wsi == 'Xcb' or self.wsi == 'Xlib' or self.wsi == 'Wayland' or self.wsi == 'Mir' or self.wsi == 'Display':
             instance_extensions=[('msg_callback_get_proc_addr', []),
                                   ('wsi_enabled',
                                   ['vkDestroySurfaceKHR',
```
Comment (pbantolas) [inline vk-layer-generate.py]: @jon-lunarg If we are to unconditionally support the surface extension (and remove Display as a wsi option), should I move all the extension-specific entry points outside the conditionals?

Comment (jon-lunarg) [inline vk-layer-generate.py]: The vkCreateDisplayPlaneSurfaceKHR would go into the "instance_extensions" variable for all the cases" Win32, Android, Xcb/Xlib/...
There probably is a better to assign to "instance_extensions" the common entry points outside the ifs so there is not so much repeated code in all the ifs.  Python is not my strong point.

**Comment by jon-lunarg** [comment]
It would be nice if all the functions for the entire VK_KHR_display  extension was added by this MR rather than just the CreateDisplayPlaneSurface  function

**Comment by pbantolas** [comment]
@jon-lunarg Let me know if I missed something.

File: layers/object_tracker.h [hunk 5]
```diff
@@ -146,7 +146,20 @@ static void createInstanceRegisterExtensions(const VkInstanceCreateInfo *pCreate
...
+    pDisp->GetDisplayPlaneSupportedDisplaysKHR = 
+        (PFN_vkGetDisplayPlaneSupportedDisplaysKHR)gpa(instance, "vkGetDisplayPlaneSupportedDisplaysKHR");
+    pDisp->GetDisplayModePropertiesKHR = 
+        (PFN_vkGetDisplayModePropertiesKHR)gpa(instance, "vkGetDisplayModePropertiesKHR");
+    pDisp->CreateDisplayModeKHR = 
+        (PFN_vkCreateDisplayModeKHR)gpa(instance, "vkCreateDisplayModeKHR");
+    pDisp->GetDisplayPlaneCapabilitiesKHR = 
+        (PFN_vkGetDisplayPlaneCapabilitiesKHR)gpa(instance, "vkGetDisplayPlaneCapabilitiesKHR");
+    pDisp->CreateDisplayPlaneSurfaceKHR = 
+        (PFN_vkCreateDisplayPlaneSurfaceKHR)gpa(instance, "vkCreateDisplayPlaneSurfaceKHR");
```
Comment (jon-lunarg) [inline layers/object_tracker.h]: I see object_tracker.h is currently problematic in that it uses one "wsi_enabled" variable for all the wsi instance extensions.  Since this is problematic for all the extensions KHR_surface, KHR_win32_surface, KHR_xlib_surface, etc.  in addition to KHR_display. I will make a note to fix this later after your Pull Request go in.  It looks like swapchain.cpp does it correctly in it's createInstanceRegisterExtensions  function.

File: layers/swapchain.cpp [hunk 6]
```diff
@@ -903,20 +920,244 @@ VKAPI_ATTR VkBool32 VKAPI_CALL GetPhysicalDeviceXlibPresentationSupportKHR(VkPhy
...
+
+    if (!skipCall) {
+        result = my_data->instance_dispatch_table->GetPhysicalDeviceDisplayPropertiesKHR(physicalDevice, pPropertyCount, pProperties);
+    }
+    return result;
+}
+
+static uint32_t gDisplayPlanePropertyCount = 0;
+static bool gDisplayPlanePropertyCountInit = false;
+VKAPI_ATTR VkResult VKAPI_CALL
```
Comment (jon-lunarg) [inline layers/swapchain.cpp]: These shouldn't be global variables as they are per physicalDevice values. They need to be added to the "struct SwpPhysicalDevice" in swapchain.h  and then accessed by finding the current physical device in the layer_data's 
 std::unordered_map<void *, SwpPhysicalDevice> physicalDeviceMap;

File: layers/swapchain.cpp [hunk 7]
```diff
@@ -903,20 +920,244 @@ VKAPI_ATTR VkBool32 VKAPI_CALL GetPhysicalDeviceXlibPresentationSupportKHR(VkPhy
...
+    lock.unlock();
+
+    if (!skipCall) {
+        result = my_data->instance_dispatch_table->GetPhysicalDeviceDisplayPlanePropertiesKHR(physicalDevice, pPropertyCount, pProperties);
+
+        if (!gDisplayPlanePropertyCountInit)
+        {
+            gDisplayPlanePropertyCount = *pPropertyCount;
+            gDisplayPlanePropertyCountInit = true;
+        }
```
Comment (jon-lunarg) [inline layers/swapchain.cpp]: I think you will need to lock around the update to these DIsplayPlanePropertyCount variables when accessing them within the layer_data map.

File: layers/swapchain.cpp [hunk 8]
```diff
@@ -903,20 +920,244 @@ VKAPI_ATTR VkBool32 VKAPI_CALL GetPhysicalDeviceXlibPresentationSupportKHR(VkPhy
...
+
+    if (!pPropertyCount) {
+        skipCall |= LOG_ERROR_NULL_POINTER(VK_DEBUG_REPORT_OBJECT_TYPE_INSTANCE_EXT, pPhysicalDevice->pInstance, "pPropertyCount");
+    }
+    lock.unlock();
+
+    if (!skipCall) {
+        result = my_data->instance_dispatch_table->GetPhysicalDeviceDisplayPropertiesKHR(physicalDevice, pPropertyCount, pProperties);
+    }
+    return result;
```
Comment (jon-lunarg) [inline layers/swapchain.cpp]: I think this need to return  VK_ERROR_VALIDATION_FAILED_EXT if you detect an error

File: layers/swapchain.cpp [hunk 9]
```diff
@@ -903,20 +920,244 @@ VKAPI_ATTR VkBool32 VKAPI_CALL GetPhysicalDeviceXlibPresentationSupportKHR(VkPhy
...
+        result = my_data->instance_dispatch_table->GetPhysicalDeviceDisplayPlanePropertiesKHR(physicalDevice, pPropertyCount, pProperties);
+
+        if (!gDisplayPlanePropertyCountInit)
+        {
+            gDisplayPlanePropertyCount = *pPropertyCount;
+            gDisplayPlanePropertyCountInit = true;
+        }
+    }
+    return result;
+}
```
Comment (jon-lunarg) [inline layers/swapchain.cpp]: I think this need to return  VK_ERROR_VALIDATION_FAILED_EXT if you detect an error

File: layers/swapchain.cpp [hunk 10]
```diff
@@ -903,20 +920,244 @@ VKAPI_ATTR VkBool32 VKAPI_CALL GetPhysicalDeviceXlibPresentationSupportKHR(VkPhy
...
+                __FUNCTION__,
+                "planeIndex",
+                gDisplayPlanePropertyCount - 1);
+    }
+    lock.unlock();
+
+    if (!skipCall) {
+        result = my_data->instance_dispatch_table->GetDisplayPlaneSupportedDisplaysKHR(physicalDevice, planeIndex, pDisplayCount, pDisplays);
+    }
+    return result;
```
Comment (jon-lunarg) [inline layers/swapchain.cpp]: I think this need to return  VK_ERROR_VALIDATION_FAILED_EXT if you detect an error

File: layers/swapchain.cpp [hunk 11]
```diff
@@ -903,20 +920,244 @@ VKAPI_ATTR VkBool32 VKAPI_CALL GetPhysicalDeviceXlibPresentationSupportKHR(VkPhy
...
+
+    if (!pPropertyCount) {
+        skipCall |= LOG_ERROR_NULL_POINTER(VK_DEBUG_REPORT_OBJECT_TYPE_INSTANCE_EXT, pPhysicalDevice->pInstance, "pPropertyCount");
+    }
+    lock.unlock();
+
+    if (!skipCall) {
+        result = my_data->instance_dispatch_table->GetDisplayModePropertiesKHR(physicalDevice, display, pPropertyCount, pProperties);
+    }
+    return result;
```
Comment (jon-lunarg) [inline layers/swapchain.cpp]: I think this need to return  VK_ERROR_VALIDATION_FAILED_EXT if you detect an error

File: layers/swapchain.cpp [hunk 12]
```diff
@@ -903,20 +920,244 @@ VKAPI_ATTR VkBool32 VKAPI_CALL GetPhysicalDeviceXlibPresentationSupportKHR(VkPhy
...
+    if (!pCreateInfo) {
+        skipCall |= LOG_ERROR_NULL_POINTER(VK_DEBUG_REPORT_OBJECT_TYPE_INSTANCE_EXT, pPhysicalDevice->pInstance, "pCreateInfo");
+    }
+    lock.unlock();
+
+    if (!skipCall) {
+        result = my_data->instance_dispatch_table->CreateDisplayModeKHR(physicalDevice, display, pCreateInfo, pAllocator, pMode);
+    }
+    return result;
+}
```
Comment (jon-lunarg) [inline layers/swapchain.cpp]: I think this need to return  VK_ERROR_VALIDATION_FAILED_EXT if you detect an error

File: layers/swapchain.cpp [hunk 13]
```diff
@@ -903,20 +920,244 @@ VKAPI_ATTR VkBool32 VKAPI_CALL GetPhysicalDeviceXlibPresentationSupportKHR(VkPhy
...
+    }
+
+    if (!pCreateInfo) {
+        skipCall |= LOG_ERROR_NULL_POINTER(VK_DEBUG_REPORT_OBJECT_TYPE_INSTANCE_EXT, pPhysicalDevice->pInstance, "pCreateInfo");
+    }
+    lock.unlock();
+
+    if (!skipCall) {
+        result = my_data->instance_dispatch_table->CreateDisplayModeKHR(physicalDevice, display, pCreateInfo, pAllocator, pMode);
+    }
```
Comment (jon-lunarg) [inline layers/swapchain.cpp]: After calling down the chain, this should add the newly created *pSurface to the surfaceMap.  See CreateWin32SurfaceKHR for an example.

File: layers/swapchain.cpp [hunk 14]
```diff
@@ -903,20 +920,244 @@ VKAPI_ATTR VkBool32 VKAPI_CALL GetPhysicalDeviceXlibPresentationSupportKHR(VkPhy
...
+
+    if (!pCapabilities) {
+        skipCall |= LOG_ERROR_NULL_POINTER(VK_DEBUG_REPORT_OBJECT_TYPE_INSTANCE_EXT, pPhysicalDevice->pInstance, "pCapabilities");
+    }
+    lock.unlock();
+
+    if (!skipCall) {
+        result = my_data->instance_dispatch_table->GetDisplayPlaneCapabilitiesKHR(physicalDevice, mode, planeIndex, pCapabilities);
+    }
+    return result;
```
Comment (jon-lunarg) [inline layers/swapchain.cpp]: I think this need to return VK_ERROR_VALIDATION_FAILED_EXT if you detect an error

File: layers/swapchain.cpp (hunk 13 above)
Comment (jon-lunarg) [inline layers/swapchain.cpp]: Sorry I added this comment to the wrong place, never mind!

File: layers/swapchain.cpp [hunk 15]
```diff
@@ -949,10 +1190,10 @@ CreateDisplayPlaneSurfaceKHR(VkInstance instance, const VkDisplaySurfaceCreateIn
             // Point to the associated SwpInstance:
             pInstance->surfaces[*pSurface] = &my_data->surfaceMap[*pSurface];
         }
-        loader_platform_thread_unlock_mutex(&globalLock);
+        lock.unlock();
         return result;
     }
-    loader_platform_thread_unlock_mutex(&globalLock);
+    lock.unlock();
     return VK_ERROR_VALIDATION_FAILED_EXT;
```
Comment (jon-lunarg) [inline layers/swapchain.cpp]: not strictly needed

File: vk-layer-generate.py [hunk 16]
```diff
@@ -1149,6 +1150,7 @@ def generate_intercept(self, proto, qual):
                                    'AcquireNextImageKHR' : ['fence', 'semaphore' ],
                                    'UpdateDescriptorSets' : ['pTexelBufferView'],
                                    'CreateSwapchainKHR' : ['oldSwapchain'],
+                                   'CreateDisplayModeKHR': ['pMode'],
                                   }
```
Comment (jon-lunarg) [inline vk-layer-generate.py]: I think this is only for objects that can be NULL. AFAIK DisplayN=Mode object can't be NULL.

File: vk-layer-generate.py [hunk 17]
```diff
@@ -1215,7 +1217,7 @@ def generate_intercept(self, proto, qual):
                      '}' % (qual, decl, proto.c_call()))
             return "".join(funcs)
         # Temporarily prevent  DestroySurface call from being generated until WSI layer support is fleshed out
-        elif 'DestroyInstance' in proto.name or 'DestroyDevice' in proto.name:
+        elif 'DestroyInstance' in proto.name or 'DestroyDevice' in proto.name or 'CreateDisplayModeKHR' in proto.name:
             return ""
```
Comment (jon-lunarg) [inline vk-layer-generate.py]: Why is this needed?

File: vk-layer-generate.py [hunk 18]
```diff
@@ -1617,33 +1610,23 @@ def generate_body(self):
...
             instance_extensions=[('wsi_enabled',
-                                  ['vkDestroySurfaceKHR',
-                                   'vkGetPhysicalDeviceSurfaceSupportKHR',
-                                   'vkGetPhysicalDeviceSurfaceCapabilitiesKHR',
-                                   'vkGetPhysicalDeviceSurfaceFormatsKHR',
-                                   'vkGetPhysicalDeviceSurfacePresentModesKHR',
-                                   'vkCreateDisplayPlaneSurfaceKHR',
-                                   'vkCreateXcbSurfaceKHR',
+                                  ['vkCreateXcbSurfaceKHR',
                                    'vkCreateXlibSurfaceKHR',
```
Comment (jon-lunarg) [inline vk-layer-generate.py]: All these changes from 1308 to 1692 are very problematic.  "instance_extensions " list is for instance exts. "extensions" list is for device exts.  You are adding WSI instance extensions into the device ext list so then they show up
in the object tracker's GetDeviceProcAddr when they need to show up in the GetInstanceProcAddr.

Addtionally you should add all seven of the KHR_display extensions  to the instance_extensions list in all cases. You are only adding the CreateDdisplayPlaneSurface.

File: vulkan.py [hunk 19]
```diff
@@ -1051,6 +1051,48 @@ def __repr__(self):
...
              Param("uint32_t*", "pPresentModeCount"),
              Param("VkPresentModeKHR*", "pPresentModes")]),
+    ],
+)
+
+ext_khr_display = Extension(
+    name="VK_KHR_display",
+    headers=["vulkan/vulkan.h"],
+    objects=[],
+    protos=[
```
Comment (jon-lunarg) [inline vulkan.py]: objects should include "VkSurfaceKHR"  and "VKDisplayModeKHR"

File: vulkan.py [hunk 20]
```diff
@@ -1245,8 +1287,8 @@ def __repr__(self):
         extensions = [core, ext_khr_surface, ext_khr_device_swapchain, ext_khr_android_surface]
         extensions_all = [core, ext_khr_surface, ext_khr_device_swapchain, ext_khr_android_surface, lunarg_debug_report]
     elif sys.argv[1] == 'Xcb' or sys.argv[1] == 'Xlib' or sys.argv[1] == 'Wayland' or sys.argv[1] == 'Mir' or sys.argv[1] == 'Display':
-        extensions = [core, ext_khr_surface, ext_khr_device_swapchain, ext_khr_xcb_surface, ext_khr_xlib_surface, ext_khr_wayland_surface, ext_khr_mir_surface]
-        extensions_all = [core, ext_khr_surface, ext_khr_device_swapchain, ext_khr_xcb_surface, ext_khr_xlib_surface, ext_khr_wayland_surface, ext_khr_mir_surface, lunarg_debug_report]
+        extensions = [core, ext_khr_surface, ext_khr_device_swapchain, ext_khr_xcb_surface, ext_khr_xlib_surface, ext_khr_wayland_surface, ext_khr_mir_surface, ext_khr_display]
+        extensions_all = [core, ext_khr_surface, ext_khr_device_swapchain, ext_khr_xcb_surface, ext_khr_xlib_surface, ext_khr_wayland_surface, ext_khr_mir_surface, ext_khr_display, lunarg_debug_report]
     else:
```
Comment (jon-lunarg) [inline vulkan.py]: Above you have added the ext_khr_display to the linux targets.  Can you also add them to the Win32 targets. I am not sure if they should be added to Android targets or not.

**Comment by jon-lunarg** [comment]
Petros:
Thanks for the updated PR!  See my specific comments for issues I found.

**Comment by pbantolas** [comment]
Jon, thanks for taking the time to go through my commit and comment with your suggestions. I'm currently on holiday and will be returning on Wednesday, so I'll get to it then!





On Tue, May 24, 2016 at 2:52 PM -0700, "Jon Ashburn" <notifications@github.com<mailto:notifications@github.com>> wrote:


Petros:
Thanks for the updated PR! See my specific comments for issues I found.

File: vk-layer-generate.py (hunk 17 above)
Comment (peterfsat) [inline vk-layer-generate.py]: There is no mapping from a VkDisplayModeKHR object to the VkDebugReportObjectTypeEXT enum, so autogeneration fails for this function. I guess it needs to be explicitly tracked, but I will need some help with that.

**Comment by jon-lunarg** [comment]
Petros:

I am taking a crack at adding on additional commits on top of your latest to fill out the object_tracker, unique_objects  and swapchain layer support of all of VK_KHR_display extension.  I likely will not get this done (due to other things I need to finish) until after I return from vacation next week.

I could share my WIP patch with you if you might be able to carry the ball forward in the meantime?

**Comment by pbantolas** [comment]
Jon, I'm also OOO until tomorrow, and I'll be back on Friday but can't make any promises! You can share it with me and I might be able to take a look next week too.


_____________________________

**Comment by jon-lunarg** [comment]
This PR is being replaced by #656 which includes these changes plus more changes required to finish support for KHR_display  extension
